### ISSUE TITLE
AttributeError: 'DropTable' object has no attribute 'if_exists'

### ISSUE THREAD

**Issue by sharthZ23** [xzkostyan/clickhouse-sqlalchemy#22]
subj

Code to reproduce:
```python
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine('clickhouse://.../testing')
df = pd.DataFrame({'value': [0, 1], 'str': ['qwer', 'asdf']})

conn = engine.connect()
df.to_sql('test_drop', conn, if_exists='replace', index=False)
conn.close()
```

Clickhouse:
```sql
CREATE TABLE testing.test_drop (value Int32, str String) ENGINE = TinyLog;
```

Full exception:
```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-20-56f4dbb057f1> in <module>()
      1 with DBConnection('ch_credentials.pkl', dbname='testing') as conn:
----> 2     test_df.to_sql('test_drop', conn, if_exists='replace', index=False)

/usr/local/lib/python3.6/site-packages/pandas/core/generic.py in to_sql(self, name, con, flavor, schema, if_exists, index, index_label, chunksize, dtype)
   1532         sql.to_sql(self, name, con, flavor=flavor, schema=schema,
   1533                    if_exists=if_exists, index=index, index_label=index_label,
-> 1534                    chunksize=chunksize, dtype=dtype)
   1535 
   1536     def to_pickle(self, path, compression='infer',

/usr/local/lib/python3.6/site-packages/pandas/io/sql.py in to_sql(frame, name, con, flavor, schema, if_exists, index, index_label, chunksize, dtype)
    471     pandas_sql.to_sql(frame, name, if_exists=if_exists, index=index,
    472                       index_label=index_label, schema=schema,
--> 473                       chunksize=chunksize, dtype=dtype)
    474 
    475 

/usr/local/lib/python3.6/site-packages/pandas/io/sql.py in to_sql(self, frame, name, if_exists, index, index_label, schema, chunksize, dtype)
   1153                          if_exists=if_exists, index_label=index_label,
   1154                          schema=schema, dtype=dtype)
-> 1155         table.create()
   1156         table.insert(chunksize)
   1157         if (not name.isdigit() and not name.islower()):

/usr/local/lib/python3.6/site-packages/pandas/io/sql.py in create(self)
    592                 raise ValueError("Table '%s' already exists." % self.name)
    593             elif self.if_exists == 'replace':
--> 594                 self.pd_sql.drop_table(self.name, self.schema)
    595                 self._execute_create()
    596             elif self.if_exists == 'append':

/usr/local/lib/python3.6/site-packages/pandas/io/sql.py in drop_table(self, table_name, schema)
   1203         if self.has_table(table_name, schema):
   1204             self.meta.reflect(only=[table_name], schema=schema)
-> 1205             self.get_table(table_name, schema).drop()
   1206             self.meta.clear()
   1207 

/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/schema.py in drop(self, bind, checkfirst)
    792         bind._run_visitor(ddl.SchemaDropper,
    793                           self,
--> 794                           checkfirst=checkfirst)
    795 
    796     def tometadata(self, metadata, schema=RETAIN_SCHEMA,

/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py in _run_visitor(self, visitorcallable, element, **kwargs)
   1547     def _run_visitor(self, visitorcallable, element, **kwargs):
   1548         visitorcallable(self.dialect, self,
-> 1549                         **kwargs).traverse_single(element)
   1550 
   1551 

/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/visitors.py in traverse_single(self, obj, **kw)
    119             meth = getattr(v, "visit_%s" % obj.__visit_name__, None)
    120             if meth:
--> 121                 return meth(obj, **kw)
    122 
    123     def iterate(self, obj):

/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/ddl.py in visit_table(self, table, drop_ok, _is_metadata_operation)
    953                 self.traverse_single(column.default)
    954 
--> 955         self.connection.execute(DropTable(table))
    956 
    957         table.dispatch.after_drop(

/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py in execute(self, object, *multiparams, **params)
    946             raise exc.ObjectNotExecutableError(object)
    947         else:
--> 948             return meth(self, multiparams, params)
    949 
    950     def _execute_function(self, func, multiparams, params):

/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/ddl.py in _execute_on_connection(self, connection, multiparams, params)
     66 
     67     def _execute_on_connection(self, connection, multiparams, params):
---> 68         return connection._execute_ddl(self, multiparams, params)
     69 
     70     def execute(self, bind=None, target=None):

/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py in _execute_ddl(self, ddl, multiparams, params)
   1001             dialect=dialect,
   1002             schema_translate_map=self.schema_for_object
-> 1003             if not self.schema_for_object.is_default else None)
   1004         ret = self._execute_context(
   1005             dialect,

<string> in <lambda>(self, bind, dialect, **kw)

/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/elements.py in compile(self, default, bind, dialect, **kw)
    440             else:
    441                 dialect = default.StrCompileDialect()
--> 442         return self._compiler(dialect, bind=bind, **kw)
    443 
    444     def _compiler(self, dialect, **kw):

/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/ddl.py in _compiler(self, dialect, **kw)
     24         Dialect."""
     25 
---> 26         return dialect.ddl_compiler(dialect, self, **kw)
     27 
     28 

/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/compiler.py in __init__(self, dialect, statement, bind, schema_translate_map, compile_kwargs)
    217             if self.can_execute:
    218                 self.execution_options = statement._execution_options
--> 219             self.string = self.process(self.statement, **compile_kwargs)
    220 
    221     @util.deprecated("0.7", ":class:`.Compiled` objects now compile "

/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/compiler.py in process(self, obj, **kwargs)
    243 
    244     def process(self, obj, **kwargs):
--> 245         return obj._compiler_dispatch(self, **kwargs)
    246 
    247     def __str__(self):

/usr/local/lib/python3.6/site-packages/sqlalchemy/sql/visitors.py in _compiler_dispatch(self, visitor, **kw)
     79                     raise exc.UnsupportedCompilationError(visitor, cls)
     80                 else:
---> 81                     return meth(self, **kw)
     82         else:
     83             # The optimization opportunity is lost for this case because the

~/Projects/Git/clickhouse-sqlalchemy/clickhouse_sqlalchemy/drivers/base.py in visit_drop_table(self, drop)
    293         text = '\nDROP TABLE '
    294 
--> 295         if drop.if_exists:
    296             text += 'IF EXISTS '
    297 

AttributeError: 'DropTable' object has no attribute 'if_exists'
```

Lib version - '0.0.5', installed from master-branch.

**Comment by wuhuanyan** [comment]
I changed it like this
clickhouse_sqlalchemy/drivers/base.py
294th lines
```
    def visit_drop_table(self, drop):
        text = '\nDROP TABLE IF EXISTS '
        # if drop.if_exists:
        #     text += 'IF EXISTS '
        return text + self.preparer.format_table(drop.element)
```

**Comment by wuhuanyan** [comment]
In addition to other questions
How do you pass ENGINE parameters of CREATE TABLE through pandas.dataframe.to_sql?

**Comment by xzkostyan** [comment]
Hi, @wuhuanyan 

`if_exists` was designed as extension of [`table.drop` syntax](https://github.com/xzkostyan/clickhouse-sqlalchemy/blob/7fe83865ef0139dc44fca0b6d0f16b02e1644adf/clickhouse_sqlalchemy/sql/schema.py#L8).

When SA reflect table from database (see query in clickhouse log `DESCRIBE TABLE test_drop FORMAT TabSeparatedWithNamesAndTypes`) it constructs instance of default Table class. That's why `drop.if_exists` in `visit_drop_table` doesn't work.

If we comment or change this string to something like this: `if getattr(drop, 'if_exists', False)`. We will get following error:

```python
~/clickhouse-sqlalchemy/clickhouse_sqlalchemy/drivers/base.py in post_create_table(self, table)
    286 
    287         if not engine:
--> 288             raise exc.CompileError("No engine for table '%s'" % table.name)
    289 
    290         return ' ENGINE = ' + self.process(engine)

CompileError: No engine for table 'test_drop'
```

Engine is required parameter for table creation. I think it also should be reflected from database via parsing `SHOW CREATE TABLE testing` query.

In current pandas [to_sql interface](https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.DataFrame.to_sql.html) it seems impossible to pass additional parameters such as CH Engine.

I'd recommend you not to drop/create table as temporary solution. Or you can drop/create tables via sqlalchemy. See [README](https://github.com/xzkostyan/clickhouse-sqlalchemy#basic-ddl-support) for examples.

**Comment by xzkostyan** [comment]
Fixed in #94.
